requests/models.py

```diff
@@ -155,8 +155,12 @@ def _encode_files(files, data):
 
             if isinstance(fp, (str, bytes, bytearray)):
                 fdata = fp
-            else:
+            elif hasattr(fp, 'read'):
                 fdata = fp.read()
+            elif fp is None:
+                continue
+            else:
```

Comment:
In this case, are we expecting the objects `__str__` representation to be sufficient? We currently blow up on things like `datetime` or `dict`, but this may have unexpected results.

If we're OK with the above, would it make more sense to reorg the conditionals to check for `read`, `None` and then everything else would be `fdata = fp`?

e.g.
```python
if hasattr(fp, 'read'):
    fdata = fp.read()
elif fp is None:
    continue
else:
    fdata = fp
```

Reply:
@nateprewitt 

```python
else:
    fdata = fp

rf = RequestField(name=k, data=fdata, filename=fn, headers=fh)
```

Getting rid of else part will result in fdata being an unbound variable. Should we set it to fdata=None in else and let it blow up or leave things as is?

Reply:
Urgh, yep. Ok so @sigmavirus24 we may just want to put the exception in now. Non-file objects were already exploding on the read call so they never worked.

We *could* make sure the exception inherits from AttributeError if we want to ensure we’re not breaking anyone currently relying on this defect for flow control.

Reply:
@nateprewitt I'd rather never inherit from `AttributeError` if possible. That is just a world of pain. We've done this in the past, but that means if people are putting too much code in their `try/except` block, we may cause them to enter an exception handling block.

In short, I think the code as is may be safe right now. If we want to be extra magical, we could do `fdata = str(fp)` but I really don't want to add that behaviour at all. Let's just leave this as is, I guess and plan to add the exception in 3.0

Reply:
Sounds good! I think I misread your first response, so we should be set then.